### PR TITLE
Product taxon position don't affect parent taxon

### PR DESCRIPTION
**Pull request by everwhatever**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.11 |
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | no |
| Deprecations?   | no |
| Related tickets | [github](https://github.com/Sylius/Sylius/issues/14308) |
| License         | MIT|

### CODE REVIEW

**Review by coldic3 — CHANGES_REQUESTED** [review]
Could you add tests to see if your fix resolves the problem described in https://github.com/Sylius/Sylius/issues/14308? You can start with behat scenario suggested by @ehibes: https://github.com/ehibes/Sylius/commit/d0608afc0d95075149f6eba16623314a641f3e2d

File: features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        And this product is in "Toys" taxon at 1st position
+        And this product is in "Soft Toys" taxon at 3rd position
+        And the store has a product "Small pug"
+        And this product is in "Toys" taxon at 3rd position
+        And this product is in "Soft Toys" taxon at 1st position
+        And the store has a product "Young pug"
+        And this product is in "Toys" taxon at 2nd position
+        And this product is in "Soft Toys" taxon at 2nd position
+
+    @ui
```
Comment (coldic3) [inline features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: Can we test it on API? It's shop API so it should be covered IMO.

Comment (TheMilek) [reply features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature:21]: Yes, I can confirm We've got it in API 👌🏽

Comment (TheMilek) [inline features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]:
```suggestion
    @api @ui
```
We need to have it covered also in API as We've got such a functionality

**Review by TheMilek — CHANGES_REQUESTED** [review]

File: features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature [hunk 2]
```diff
@@ -0,0 +1,28 @@
+@viewing_products
+Feature: Viewing products from specific taxon sorted by position
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: Change the title to be consistent with the filename

File: features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature [hunk 3]
```diff
@@ -0,0 +1,28 @@
+@viewing_products
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: I would be for merging this feature file with `viewing_product_from_specific_taxon_sorted_by_position.feature` and check sorted products by browsing two different taxons (parent and children)

File: features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        And this product is in "Soft Toys" taxon at 2nd position
+
+    @ui
+    Scenario: Seeing sorted product on list from a specific taxon
+        When I browse products from taxon "Toys"
+        Then I should see 3 products in the list
+        And they should have order like "Old pug", "Young pug" and "Small pug"
+        When I browse products from taxon "Soft Toys"
+        Then I should see 3 products in the list
+        And they should have order like "Small pug", "Young pug" and "Old pug"
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: It should be extracted to the second scenario

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 5]
```diff
@@ -85,6 +84,12 @@ public function createShopListQueryBuilder(
             ;
         }
 
+        if (empty($sorting)){
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]:
```suggestion
        if (empty($sorting)) {
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 6]
```diff
@@ -85,6 +84,12 @@ public function createShopListQueryBuilder(
...
+            $queryBuilder->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
+                ->orderBy('productTaxons.position', 'ASC')
+                ->setParameter('taxonId', $taxon->getId());
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]:
```suggestion
            $queryBuilder
                ->leftJoin('o.productTaxons', 'productTaxons', 'WITH', 'productTaxons.taxon = :taxonId')
                ->orderBy('productTaxons.position', 'ASC')
                ->setParameter('taxonId', $taxon->getId())
            ;
```

**Review by GSadee — COMMENTED** [review]
Change the title to be more descriptive and add to related tickets the issue, which this PR fixes (instead of jira ticket which is not public).

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php (hunk 5 above)
Comment (ehibes) [reply src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: @everwhatever just missing a space here

**Comment by ehibes** [comment]
Behat test is failing with api context. 
IMO it is normal, since the API controller does not use the same method of the repository. 
So we have to create a ProductCollectionDataProvider for the api part to use the right query builder.

@GSadee @coldic3 Could you accept this pull request for the shop part, even if we open a new issue/PR for the api part?

**Comment by GSadee** [comment]
> Behat test is failing with api context. IMO it is normal, since the API controller does not use the same method of the repository. So we have to create a ProductCollectionDataProvider for the api part to use the right query builder.
> 
> @GSadee @coldic3 Could you accept this pull request for the shop part, even if we open a new issue/PR for the api part?

Yes, as the API part requires additional and slightly different work to fix the issue, we could do that in a separate PR 👌🏻

File: features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature [hunk 7]
```diff
@@ -6,16 +6,26 @@ Feature: Viewing products from specific taxon sorted by position
...
+        And this product is in "Toys" taxon at 1st position
         And this product is in "Soft Toys" taxon at 3rd position
+        And the store has a product "Small pug"
+        And this product is in "Toys" taxon at 3rd position
+        And this product is in "Soft Toys" taxon at 1st position
         And the store has a product "Young pug"
+        And this product is in "Toys" taxon at 2nd position
         And this product is in "Soft Toys" taxon at 2nd position
 
-    @ui @api
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:17]: Does it not work in both scenarios with the `@api` tag?

File: features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature [hunk 8]
```diff
@@ -6,16 +6,26 @@ Feature: Viewing products from specific taxon sorted by position
...
+    @ui
     Scenario: Seeing sorted product on list from a specific taxon
-        When I browse products from taxon "Soft Toys"
+        When I browse products from taxon "Toys"
         Then I should see 3 products in the list
         And they should have order like "Old pug", "Young pug" and "Small pug"
+
+     @ui
+    Scenario: Seeing sorted product on list from a child taxon
```
Comment (GSadee) [inline features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature]:
```suggestion
    @ui
    Scenario: Seeing sorted product on list from a child taxon
```

File: features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature (hunk 1 above)
Comment (everwhatever) [reply features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: in other PR

Comment (everwhatever) [reply features/product/viewing_products/viewing_product_from_parent_level_taxon_sorted_by_position.feature]: in other PR

File: features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature (hunk 7 above)
Comment (everwhatever) [reply features/product/viewing_products/viewing_product_from_specific_taxon_sorted_by_position.feature:17]: only with ui

**Comment by Rafikooo** [comment]
It would be nice if you tidy up your commits better, for instance, squash or reset all commits `git reset HEAD~8` and then commit them again with a more descriptive commit message

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, @everwhatever! :tada:

**Pull request merged**